libs/auth-nest-tools/src/lib/auth.middleware.ts

```
@@ -18,22 +19,111 @@ interface Middleware {
+  scope: string
+}
+
 /**
  * Middleware that adds user authorization and information to OpenAPI Client requests.
  */
 export class AuthMiddleware implements Middleware {
-  constructor(private auth: Auth, private forwardUserInfo: boolean = true) {}
+  constructor(
+    private auth: Auth,
+    private options: AuthMiddlewareOptions = {
+      forwardUserInfo: true,
+    },
+    private logger?: Logger,
+  ) {}
 
   async pre(context: RequestContext) {
+    let bearerToken = this.auth.authorization
+
+    if (this.options.tokenExchangeOptions) {
+      const accessToken = await this.exchangeToken(
+        bearerToken.replace('Bearer ', ''),
+        this.options.tokenExchangeOptions,
+      )
+
+      bearerToken = `Bearer ${accessToken}`
+    }
+
     context.init.headers = Object.assign({}, context.init.headers, {
-      authorization: this.auth.authorization,
+      authorization: bearerToken,
     })
 
-    if (this.forwardUserInfo) {
+    if (this.options.forwardUserInfo) {
       context.init.headers = Object.assign({}, context.init.headers, {
         'User-Agent': this.auth.userAgent,
         'X-Real-IP': this.auth.ip,
       })
     }
   }
+
+  private async exchangeToken(
+    accessToken: string,
+    options: TokenExchangeOptions,
+  ): Promise<string> {
+    try {
+      const response = await fetch(`${options.issuer}/connect/token`, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+        },
+        body: new URLSearchParams({
+          client_id: options.clientId,
+          grant_type: 'urn:ietf:params:oauth:grant-type:token-exchange',
+          scope: options.scope,
+          client_secret: options.clientSecret,
+          subject_token: accessToken,
+          subject_token_type: 'urn:ietf:params:oauth:token-type:access_token',
+        }),
+      })
+
+      if (!response.ok) {
+        this.logError(`Token exchange failed, ${await response.text()}`)
+        return ''
+      }
```

Should throw an error instead if token exchange fails?

---

Might be good to include the requested URL from the context in the error/log.